sdk/eventhub/Azure.Messaging.EventHubs.Processor/tests/Processor/EventProcessorClientLiveTests.cs

```diff
@@ -31,5 +36,831 @@ namespace Azure.Messaging.EventHubs.Tests
+                {
+                    await producer.SendAsync(batch, cancellationSource.Token);
+
+                    sentCount += batch.Count;
+                    batch.Dispose();
+                }
+            }
+
+            Assert.That(sentCount, Is.EqualTo(sourceEvents.Count), "Not all of the source events were sent.");
+
+            // Attempt to read back the events.
+
+            var completionSource = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
+            var options = new EventProcessorOptions { LoadBalancingUpdateInterval = TimeSpan.FromMilliseconds(250) };
+            var processor = CreateProcessor(scope.ConsumerGroups.First(), connectionString, options: options);
+
+            processor.ProcessErrorAsync += args =>
+            {
+                Assert.Fail($"Processor Error Surfaced: ({ args.Exception.GetType().Name })[{ args.Exception.Message }]");
+                return Task.CompletedTask;
+            };
+
+            processor.ProcessEventAsync += args =>
+            {
+                if (args.HasEvent)
+                {
+                    var eventId = args.Data.Properties[CustomIdProperty].ToString();
+
+                    // Guard against duplicates; Event Hubs has an at-least-once guarantee.
+
+                    if ((TryAdd(processedEvents, eventId, args.Data)) && (Interlocked.Increment(ref processCount) >= sentCount))
+                    {
+                        completionSource.TrySetResult(true);
+                    }
+                }
+
+                return Task.CompletedTask;
+            };
+
+            await processor.StartProcessingAsync(cancellationSource.Token);
+
+            await Task.WhenAny(completionSource.Task, Task.Delay(Timeout.Infinite, cancellationSource.Token));
+            Assert.That(cancellationSource.IsCancellationRequested, Is.False, "The cancellation token should not have been signaled.");
+
+            await processor.StopProcessingAsync(cancellationSource.Token);
+            cancellationSource.Cancel();
+
+            // Validate the events that were processed.
+
+            foreach (var sourceEvent in sourceEvents)
+            {
+                var sourceId = sourceEvent.Properties[CustomIdProperty].ToString();
+
+                if (!processedEvents.TryGetValue(sourceId, out var processedEvent))
+                {
+                    Assert.Fail($"The event with custom identifier [{ sourceId }] was not processed.");
+                }
+
+                Assert.That(sourceEvent.IsEquivalentTo(processedEvent), $"The event with custom identifier [{ sourceId }] did not match the corresponding processed event.");
+            }
+        }
+
+        /// <summary>
+        ///   Verifies that the <see cref="EventProcessorClient" /> can read a set of published events.
+        /// </summary>
+        ///
+        [Test]
+        public async Task EventsCanBeReadByOneProcessorClientUsingAnIdentityCredential()
```

Comment:
General observation, perhaps for a later issue; It seems that most of the setup code for these tests could be de-duped into a helper method so that the differentiating setups or assertions are easier to pick out per test.

Reply:
Some could be, some not so much.   In general, though they do feel a bit cluttered and have similarities, I didn't feel that there was a clear and easy way to abstract more without falling a bit deeper into the magic black box territory.   

If you've got some ideas, please feel free to give it a go;  you may be seeing some things that I'm overlooking.